### DISCUSSION TITLE
EJS Templates, Environment Variables and Cache Invalidation

### DISCUSSION THREAD

**Discussion by sailerinteractive** [parcel-bundler/parcel#8573] [General]
Consider the following EJS file index.ejs. 

```
<!doctype html>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Cache Test</title>
  </head>
  <body>
    <h1><%= process.env.SITE_LANG %></h1> 
  </body>
</html>   
```

Then i run parcel with:

```
SITE_LANG=EN npx parcel index.ejs 
```

When i run parcel again with a different value of SITE_LANG the transformer cache is not invalidated and the change is not reflected in the final output. What do i need to to do in order to have the cache invalidated when the environment variable changes. Is this something that needs to be implemented in the transformer plugin?

Im using parcel 2.7.0 and parcel-transformer-ejs 1.0.0

Thanks a lot!

**Comment by mischnic** [comment]
In the `transform` method, parcel-transformer-ejs needs to call `asset.invalidateOnEnvChange('SITE_LANG');` for any env variable that should cause invalidation.

Some examples of other transformers:
https://github.com/parcel-bundler/parcel/blob/59610c14afc1bb0fe30c00374e9f2c8bbe42463a/packages/transformers/elm/src/ElmTransformer.js#L38

https://github.com/parcel-bundler/parcel/blob/59610c14afc1bb0fe30c00374e9f2c8bbe42463a/packages/transformers/js/src/JSTransformer.js#L534-L536

**Reply by sailerinteractive** [reply]
Awesome, i just added that line to the transformer and it works like a charm. In order to implement this in a more generic way, should a transformer parse the input in order to find all used environment variables? Or rather add a config option that allows users to specify which vars should be tracked? I could try to contribute this to parcel-transformer-ejs...

**Reply by mischnic** [reply]
You should talk to the maintainer of that transformer (open an issue for this repo https://github.com/sarscode/parcel-transformer-ejs). From a user perspective it would ideally find all used env vars automatically. But that depends on what information the EJS library exposes.

**Reply by sailerinteractive** [reply]
Will do, thanks for your quick reply!
